lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+-- also a user-specified output with the same payment key as the pointer output,
+-- `applyTx` will track the first one as the new pointer.
+data Tx = Tx
+    { certs :: [Cert]
+    , inputs :: [(TxIn, Coin)]
+    , outputs :: [TxOut]
+    }
+    deriving (Eq, Generic)
+    deriving Show via (Quiet Tx)
+
+instance Semigroup Tx where
+    (Tx cs1 is1 os1) <> (Tx cs2 is2 os2) =
+        Tx (cs1 <> cs2) (is1 <> is2) (os1 <> os2)
+
+data Cert
+    = RegisterKey RewardAccount
+    | Delegate RewardAccount
+      -- ^ Which pool we're delegating to is here (and for now) irrelevant.
+      -- The main thing is that there exists a witness on-chain for this stake
+      -- key (registration certs don't require witnesses)
+      --
+      -- TODO: We may also want to add the PoolId here.
+    | DeRegisterKey RewardAccount
+    deriving (Eq, Show, Generic)
+
+-- | Given a `DelegationState`, produce a `Tx` registering or de-registering
+-- stake-keys, in order to have @n@ stake-keys.
+--
+-- E.g. @setPortfolioOf s0 _ _ 2@ creates a tx which after application causes
+-- the state to have @activeKeys == [0,1]@
+--
+-- Returns @Nothing@ if the target @n@ is already reached.
+setPortfolioOf
+    :: (SoftDerivation k, ToRewardAccount k)
+    => DelegationState k
+    -> Coin
+        -- ^ minUTxOVal
+    -> (k 'AddressK XPub -> Address)
+        -- ^ A way to construct an Address
+    -> (RewardAccount -> Bool)
+        -- ^ Whether or not the key is registered.
+        --
+        -- TODO: Need a Set or Map for the real implementation with LSQ.
+    -> Int
+        -- ^ Target number of stake keys.
+    -> Maybe Tx
+setPortfolioOf s minUTxOVal mkAddress isReg n =
+    repairKey0IfNeededTx <> changeStateTx
+  where
+    repairKey0IfNeededTx :: Maybe Tx
+    repairKey0IfNeededTx = case repairKey0 $ state s of
+        [] -> Nothing
+        cs -> Just $ Tx cs [] []
+      where
+        repairKey0 (More _ _ False) | n > 0 = deleg [minBound]
+        repairKey0 _ = []
+
+    changeStateTx :: Maybe Tx
+    changeStateTx = txWithCerts $ case compare (toEnum n) (nextKeyIx s) of
+        GT -> deleg [nextKeyIx s .. toEnum (n - 1)]
+        EQ -> []
+        LT -> dereg $ reverse [toEnum n .. (pred $ nextKeyIx s)]
+
+      where
+        txWithCerts [] = Nothing
+        txWithCerts cs = Just $ Tx
+            { certs = cs
+            , inputs = maybeToList (mkTxIn <$> pointer s)
+            , outputs = maybeToList $
+                -- Note that this is the only place where we move the pointer.
+                -- I.e. repairKey0IfNeededTx won't do it on its own.
+                (\i -> TxOut
+                    (mkAddress $ keyAtIx s i)
+                    (TB.fromCoin minUTxOVal)
+                ) <$> pointerIx n
+            }
+          where
+            mkTxIn (PointerUTxO i c) = (i, c)
+            -- Note: If c > minUTxOVal we need to rely on the wallet to return the
+            -- difference to the user as change.
```

Comment:
That will be interesting to incorporate in the coin selection and is similar to some of the needs of the Plutus application backend. The wallet will need to be able to "continue" the construction of a transaction from a pre-filled one.